sympy/sets/fancysets.py

```diff
@@ -576,50 +598,62 @@ def reversed(self):
         """
         if not self:
             return self
-        return self.func(
-            self.stop - self.step, self.start - self.step, -self.step)
+        start, stop, step = self.start, self.stop, self.step
+        return self.func(stop - step, start - step, -step)
 
     def _contains(self, other):
+        from sympy.functions.elementary.piecewise import Piecewise
+        from sympy.functions.elementary.integers import ceiling
+        other = _sympify(other)
         if not self:
             return S.false
         if other.is_infinite:
             return S.false
         if not other.is_integer:
             return other.is_integer
         ref = self.start if self.start.is_finite else self.stop
-        if (ref - other) % self.step:  # off sequence
-            return S.false
-        return _sympify(other >= self.inf and other <= self.sup)
+        start, stop, step = self.start, self.stop, self.step
+        inf = Piecewise((start, Ge(step, S.Zero)), (stop - step, True))
+        sup = Piecewise((stop - step, Ge(step, S.Zero)), (start, True))
+        return Piecewise((False, Le(ceiling((stop - start)/step), S.Zero)),
+                         (False, Ne((ref - other) % self.step, S.Zero)),
+                         (True, Ge(other, inf) & Le(other, sup)),
+                         (False, True))
 
     def __iter__(self):
         if self.start in [S.NegativeInfinity, S.Infinity]:
             raise ValueError("Cannot iterate over Range with infinite start")
-        elif self:
+        if (self.size.is_Integer != True) and (self.size not in [S.Infinity, S.NegativeInfinity]):
+            raise ValueError("Cannot iterate over symbolic sized range.")
+        if self:
             i = self.start
             step = self.step
 
-            while True:
-                if (step > 0 and not (self.start <= i < self.stop)) or \
-                   (step < 0 and not (self.stop < i <= self.start)):
-                    break
+            while And(Or(Le(step, 0), And(Le(self.start, i), Lt(i, self.stop))),
+                     Or(Ge(step, 0), And(Lt(self.stop, i), Le(i, self.start)))) != False:
```

Comment:
Why is this complicated construction needed?

We have already established from above that size is either an Integer or oo (or -oo -- can the size really be -oo?). If the size is an Integer then we can just do:
```
    for n in range(size):
        yield start + n*step
```
If the size is infinite then it's the same but with itertools.count instead of range.

Reply:
The following test fails on applying the above idea. I think this happens because `range` accepts only `int` as parameter and `oo` cannot be converted to `int`.

```python
Traceback (most recent call last):
  File "/home/gagandeep/sympy/sympy/sets/tests/test_fancysets.py", line 583, in test_intersections
    for x in take(10, S.Integers.intersect(Interval(3, oo)) ))
  File "/home/gagandeep/sympy/sympy/sets/tests/test_fancysets.py", line 571, in take
    return list(itertools.islice(iterable, n))
  File "/home/gagandeep/sympy/sympy/sets/fancysets.py", line 631, in __iter__
    for n in range(int(self.size)):
  File "/home/gagandeep/sympy/sympy/core/expr.py", line 294, in __int__
    raise TypeError("can't convert %s to int" % r)
TypeError: can't convert oo to int
```
PS - Iterating through a `oo` sized `Range` is allowed in the current PR.